### PR TITLE
Automatically use WrapDB fallback

### PR DESCRIPTION
**Pull request by xclaesse**
New command `meson wrap update-db` downloads the DB into `subprojects/wrapdb.json`. If that file exists, dependency() will automatically download missing wrap files from WrapDB.

### CODE REVIEW

**Comment by tp-m** [comment]
Not that it's not a cool feature, but will there be an option for project maintainers to disable this or provide a positive-list of 'approved wraps' or at least make users jump through hoops?

e.g. as a gst-build maintainer I will then have to deal with everyone who's got any problem with any wrap that's in the db...

**Comment by xclaesse** [comment]
@tp-m It should not really affect GStreamer since we have wrap files already. It won't use WrapDB for deps already provided by a wrap. Users can already do `meson wrap install <foo>` for each dep manually, it just automate the process.

**Comment by xclaesse** [comment]
Note that you could commit your own wrapdb.json file into your project. It won't prevent users from overriding it with latest one from WrapDB (by running `meson wrap update-db`), but it could make sense for a project to commit a subset of the json with specific versions available in WrapDB, for example. When all your deps are available in WrapDB (which is a long term goal of the v2 simplification) then that could be easier to ship your wrapdb.json than each individual wrap file.

**Comment by xclaesse** [comment]
@dabrain34 the wrapdb json file is not downloaded by default, the user needs to run `meson wrap update-db` to fetch the json first. That makes this feature opt-in and not enabled by default.

If the json file is there when meson configure the project, for any missing dependency that it can find in the json, it will download the wrap file from wrapdb and use it. It is then up to the user to commit/push those downloaded .wrap files into their project if they want.

**Comment by xclaesse** [comment]
Rebased this PR, any one interested in doing a review please?

**Comment by jpakkane** [comment]
Functionality-wise this is something we do want, it makes a lot of things easier. Opt-in is probably the correct default here. There should be a way of removing the local db so that automatic resolution no longer happens.

File: docs/markdown/Using-wraptool.md [hunk 1]
```diff
@@ -82,3 +82,13 @@ straightforward:
...
+
+## Automatic dependency fallback
+
+Since *0.64.0* Meson can use WrapDB to automatically find missing dependencies.
+
+The user simply needs to download latest database:
+    $ meson wrap update-db
+
+Once the database is available locally, any dependency not found on the system
+but available in WrapDB will automatically be downloaded.
```
Comment (jpakkane) [inline docs/markdown/Using-wraptool.md:94]: The path `subprojects/wradb.json` file should be mentioned here, too. Having it only in release notes has poor discoverability.

Comment (xclaesse) [reply docs/markdown/Using-wraptool.md:95]: done

**Comment by xclaesse** [comment]
> There should be a way of removing the local db so that automatic resolution no longer happens

Not sure we need a specific subcommand just to `rm subprojects/wrapdb.json`. Note that `--wrap-mode=nodownload` also prevents downloading new wraps.

**Comment by jpakkane** [comment]
> Not sure we need a specific subcommand

I'm not saying that we do, just that there needs to be documentation on how to do it manually.

**Comment by xclaesse** [comment]
> I'm not saying that we do, just that there needs to be documentation on how to do it manually.

Ok added a sentence to make that explicit.

**Pull request merged**